templates/boltzmann/core/honeycomb.ts

```diff
@@ -0,0 +1,1097 @@
+  // you would debug level logging normally
+  verbose(message: string, ...args: unknown[]): void {
+    this._log('debug', `VERBOSE: ${message}`, args)
+  }
+
+  // Ideally we would log to a bole logger. However, logger setup happens
+  // relatively late - in the log middleware - and it's very likely that such
+  // a log won't be in-place when we stand up the middleware stack! Therefore,
+  // we do log to bole if it's set in the middleware but will fall back to
+  // good ol' fashioned JSON.stringify if need be.
+  constructor() {
+    this._stream = process.stdout
+
+    // So log messages roughly match what's output by the bole in dev mode :^)
+    if (isDev()) {
+      const pretty = require('bistre')({ time: true })
+      this._stream = pretty.pipe(this._stream)
+    }
+  }
+
+  private _log(level: 'error' | 'warn' | 'info' | 'debug', message: string, args: unknown[]): void {
+    let isSelfTest = false
+    void `{% if selftest %}`
+    isSelfTest = true
+    void `{% endif %}`
+    if (isSelfTest) {
+      return
+    }
+
+    // Log to bole if we have it
+    if (this.logger) {
+      this.logger[level](message, ...args)
+      return
+    }
+
+    const line: any = {
+      time: (new Date()).toISOString(),
+      level,
+      name: 'boltzmann:honeycomb',
+      message,
+      args
+    }
+
+    try {
+      // are the args JSON-serializable?
+      this._writeLine(JSON.stringify(line))
+      // SURVEY SAYS...
+    } catch (_) {
+      // ...ok, make it a string as a fallback
+      line.args = require('util').format('%o', line.args)
+      this._writeLine(JSON.stringify(line))
+    }
+  }
+
+  private _writeLine(line: string): void {
+      this._stream.write(Buffer.from(line + '\n'))
+  }
+
+}
+
+// We only do OpenTelemetry logging if boltzmann's main log level is debug
+const _diagLogger = new HoneycombDiagLogger()
+
+if (!process.env.LOG_LEVEL || process.env.LOG_LEVEL === 'debug') {
+  otel.diag.setLogger(_diagLogger, otelCore.getEnv().OTEL_LOG_LEVEL)
+}
+
+// There's a bug in the trace base library where the SimpleSpanProcessor doesn't
+// actually conform to the SpanProcessor interface! onStart in particular
+// doesn't take the context argument. This makes typescript extremely
+// cranky.
+//
+// We work around this by defining a new type which is the same as a
+// SimpleSpanProcessor, but without enforcing onStart and onEnd. Then, we
+// implement those methods as specified by the SpanProcessor interface.
+type _OtelSpanProcessorClass = new(_exporter: otelTraceBase.SpanExporter) => {
+  [P in Exclude<keyof otelTraceBase.SpanProcessor, 'onStart' | 'onEnd'>]: otelTraceBase.SpanProcessor[P]
+}
+
+const _OtelSpanProcessor: _OtelSpanProcessorClass = otelTraceBase.SimpleSpanProcessor
+
+class HoneycombSpanProcessor extends _OtelSpanProcessor implements otelTraceBase.SpanProcessor {
+  constructor(_exporter: otelTraceBase.SpanExporter) {
+    super(_exporter)
+  }
```

Comment:
I don't remember for *sure* but I think some of this might be more necessary than it looks due to the typing shenanigans I'm up to in this class. I'll check, and if things break I'll add some comments.

Reply:
I took a peek, and the constructor *isn't* necessary but the `onEnd` override *is*. I deleted the former and commented the latter, so we should be g2g.